src/computability/tm_to_partrec.lean

```diff
@@ -0,0 +1,885 @@
+import computability.halting
+import computability.turing_machine
+import data.num.lemmas
+
+open function (update)
+open relation
+
+namespace turing
+
+namespace to_partrec
+
+inductive code
+| zero'
+| succ
+| tail
+| cons : code → code → code
+| comp : code → code → code
+| case : code → code → code
+| fix : code → code
+
+def code.eval : code → list ℕ →. list ℕ
+| code.zero' := λ v, pure (0 :: v)
+| code.succ := λ v, pure [v.head.succ]
+| code.tail := λ v, pure v.tail
+| (code.cons f fs) := λ v, do n ← code.eval f v, ns ← code.eval fs v, pure (n.head :: ns)
+| (code.comp f g) := λ v, g.eval v >>= f.eval
+| (code.case f g) := λ v, v.head.elim (f.eval v.tail) (λ y _, g.eval (y :: v.tail))
+| (code.fix f) := pfun.fix $ λ v, (f.eval v).map $ λ v,
+  if v.head = 0 then sum.inl v.tail else sum.inr v.tail
+
+namespace code
+
+def nil : code := tail.comp succ
+def id : code := tail.comp zero'
+def head : code := cons id nil
+def zero : code := cons zero' nil
+def pred : code := case zero head
+def rfind (f : code) : code := comp pred $ comp (fix $ cons f $ cons succ tail) zero'
+def prec (f g : code) : code :=
+let c := cons tail $ cons succ $ cons (comp pred tail) $
+  cons (comp g $ cons id $ comp tail tail) $ comp tail $ comp tail tail in
+cons (comp (case id $ comp (comp (comp tail tail) (fix c)) zero')
+  (cons head $ cons (comp f tail) tail)) nil
+
+local attribute [simp] is_lawful_monad.pure_bind eval
+local attribute [simp] theorem nil_eval (v) : nil.eval v = pure [] := by simp [nil]
```

Comment:
```suggestion
@[simp] theorem nil_eval (v) : nil.eval v = pure [] := by simp [nil]
```
All of them look like reasonable simp lemmas.

Reply:
I thought it might be a good idea not to pollute the global simp set with these, since they are only used ~once. I suppose the need is minimal here since there is a sensible and unique head symbol, namely `code.eval`, which does not appear elsewhere.